Upgrade Modal step operator to support `modal>=1`

strickvl opened the pull request:
## What this does

Upgrades the ZenML Modal step operator to work with the **Modal Python SDK `>=1.0`**.

Modal's 1.0 release removed the private/internal APIs the operator previously used to build container images from a registry. This PR swaps those internals for Modal's public 1.x APIs, bumps the dependency pin, and tightens the step-operator runtime wiring around Modal auth, environments, resources, and sandbox submission.

## Changes

- **Dependency pin:** `modal>=0.64.49,<1` → `modal>=1,<2`, keeping the integration explicitly on Modal SDK v1 until v2 is tested.
- **Image building:** replaced the private pre-1.0 path —
  `modal.Image._from_args(...)` + `modal.image._ImageRegistryConfig(api_pb2.REGISTRY_AUTH_TYPE_STATIC_CREDS, ...)` + `modal.secret._Secret.from_dict(...)` —
  with the public 1.x API:
  `modal.Image.from_registry(image, secret=modal.Secret.from_dict({...}))`,
  and dropped the now-unneeded `from modal_proto import api_pb2` import.
- **Runtime environment handling:** ZenML step runtime environment variables are now passed to `modal.Sandbox.create(env=...)` instead of being attached to the Modal image definition with `Image.env(...)`, avoiding sensitive runtime values appearing in Modal image build logs. Obvious sensitive runtime credentials such as `ZENML_STORE_API_TOKEN` are split out of plain `env` and passed through `modal.Secret.from_dict(...)` / `Sandbox.create(secrets=[...])` instead.
- **Modal auth/environment wiring:** `token_id` and `token_secret` now create a lazily cached explicit `modal.Client.from_credentials(...)` client that is passed to `App.lookup`, `Sandbox.create`, and `Sandbox.from_id`. If no token pair is configured, ZenML passes `client=None` so Modal uses its normal ambient auth. `modal_environment` is passed explicitly to `App.lookup(..., environment_name=...)`; ZenML no longer temporarily mutates `MODAL_TOKEN_ID`, `MODAL_TOKEN_SECRET`, `MODAL_ENVIRONMENT`, or unsupported `MODAL_WORKSPACE` values. The unsupported `workspace` config field was removed/deferred.
- **Sandbox submission correctness:** `submit()` now preserves the entrypoint command as argv items with `modal.Sandbox.create(*entrypoint_command, ...)` instead of flattening it through `bash -c " ".join(...)`. If metadata publication or local metadata update fails after sandbox creation, ZenML now terminates the sandbox before re-raising so it does not leave an untracked remote process running.
- **Timeout handling:** the configured Modal step-operator `timeout` is now passed to `Sandbox.create()` instead of always using 24h.
- **Registry auth behavior:** registry credentials are passed to Modal when available; if none are configured, Modal attempts an anonymous image pull.
- **Resource handling:** GPU type/count validation is stricter, explicit `gpu_count=0` now means CPU-only, and memory requests are rounded up to Modal's MB integer value.
- **Cleanup/tests/docs:** removed the now-unused Modal utility path, centralized the Click `CliRunner` compatibility helper, tightened the root CLI smoke test, removed unsupported Modal workspace wording, and updated Modal docs to match the runtime behavior.

## Testing

Targeted checks run locally:

```shell
uv run pytest tests/integration/integrations/modal/flavors/test_flavor_config.py tests/integration/integrations/modal/step_operators/test_modal_step_operator.py
# 16 passed

uv run pytest tests/integration/integrations/modal/step_operators/test_modal_step_operator.py
# 10 passed

uv run pytest tests/integration/integrations/modal/step_operators/test_modal_step_operator.py tests/integration/integrations/modal/flavors/test_flavor_config.py
# 16 passed

uv run pytest tests/integration/integrations/modal/step_operators/test_modal_step_operator.py tests/integration/integrations/modal/flavors/test_flavor_config.py tests/integration/functional/cli/test_cli.py -q
# 25 passed

uv run ruff check --ignore D100,D102,D103,D107 src/zenml/integrations/modal/flavors/modal_step_operator_flavor.py src/zenml/integrations/modal/step_operators/modal_step_operator.py tests/integration/integrations/modal/flavors/test_flavor_config.py tests/integration/integrations/modal/step_operators/test_modal_step_operator.py tests/integration/functional/cli/test_model.py tests/integration/functional/cli/test_tag.py

uv run ruff format --check src/zenml/integrations/modal/flavors/modal_step_operator_flavor.py src/zenml/integrations/modal/step_operators/modal_step_operator.py tests/integration/integrations/modal/flavors/test_flavor_config.py tests/integration/integrations/modal/step_operators/test_modal_step_operator.py tests/integration/functional/cli/test_model.py tests/integration/functional/cli/test_tag.py

bash scripts/format.sh --no-yamlfix src/zenml/integrations/modal/step_operators/modal_step_operator.py tests/integration/integrations/modal/step_operators/test_modal_step_operator.py
```

Modal SDK 1.x signatures were verified against installed Modal `1.4.3` for the public APIs used here (`Image.from_registry`, `App.lookup`, `Sandbox.create`, `Sandbox.from_id`, `Sandbox.create(env=...)`, `Client.from_credentials`, and explicit `client=` / `environment_name=` wiring). The latest refactor removes temporary Modal SDK env-var mutation; non-default Modal Environment behavior should be revalidated end-to-end through `App.lookup(..., environment_name=...)` in full CI or manual stack testing.

Latest targeted checks after the explicit-client refactor:

```shell
uv run pytest tests/integration/integrations/modal/step_operators/test_modal_step_operator.py tests/integration/integrations/modal/flavors/test_flavor_config.py
# 22 passed

uv run ruff check src/zenml/integrations/modal/step_operators/modal_step_operator.py src/zenml/integrations/modal/flavors/modal_step_operator_flavor.py
# All checks passed

uv run mypy src/zenml/integrations/modal/step_operators/modal_step_operator.py
# Success: no issues found in 1 source file
```


Latest targeted checks after sandbox hardening:

```shell
uv run pytest tests/integration/integrations/modal/step_operators/test_modal_step_operator.py tests/integration/integrations/modal/flavors/test_flavor_config.py
# 25 passed, 1 warning

uv run bash scripts/lint.sh
# Ruff, pydoclint, and yamlfix passed; zizmor failed while trying to reach GitHub for actions/checkout tag lookup (TLS handshake EOF), unrelated to the three Modal files changed in the latest commit.
```


Latest follow-up after review comments:

```shell
uv run ruff format src/zenml/integrations/modal/step_operators/modal_step_operator.py tests/integration/integrations/modal/step_operators/test_modal_step_operator.py
# 2 files left unchanged

uv run ruff check src/zenml/integrations/modal/step_operators/modal_step_operator.py
# All checks passed

uv run ruff check tests/integration/integrations/modal/step_operators/test_modal_step_operator.py --extend-ignore D
# All checks passed

uv run mypy src/zenml/integrations/modal/step_operators/modal_step_operator.py
# Success: no issues found in 1 source file

uv run pydoclint src/zenml/integrations/modal/step_operators/modal_step_operator.py
# No violations

uv run pytest tests/integration/integrations/modal/step_operators/test_modal_step_operator.py tests/integration/integrations/modal/flavors/test_flavor_config.py
# 24 passed, 1 warning
```

This follow-up also fixes the CI mypy invariance error for `modal.Secret.from_dict(...)` by typing the sensitive runtime secret environment as `dict[str, str | None]`.

End-to-end remote sandbox execution should still be validated against a real Modal account in full CI or manual stack testing.

Additional manual validation:

- Ran a local smoke pipeline on the `modal` stack against the Modal `dev` environment.
- Validated one CPU-only Modal step and one GPU Modal step (`L4`, `gpu_count=1`) end-to-end.
- The GPU step ran `nvidia-smi` and checked `torch.cuda.is_available()` inside the Modal sandbox.
- Confirmed the pipeline completed successfully with both Modal step outputs flowing into the final step.
- Removed the deprecated `environment_name` argument from `Sandbox.create`; Modal environment selection now happens through `App.lookup(..., environment_name=...)`. ZenML no longer uses a temporary local `MODAL_ENVIRONMENT` override.

## Reviewer notes

The core compatibility fix is the public Modal image API replacement. The follow-up runtime changes avoid shell command flattening, keep ZenML runtime env vars out of Modal image build logs, use explicit Modal clients for configured credentials, preserve ambient auth with `client=None`, and align docs/tests with the actual Modal SDK >=1 behavior.

strickvl committed:
v0

strickvl committed:
Refactor Modal step operator GPU configuration

Inline GPU configuration logic directly into the step operator instead of
using a utility function. The logic is simple enough (10 lines of string
formatting) that a shared utility adds unnecessary indirection.

Changes:
- Inline GPU value construction in modal_step_operator.py
- Remove get_gpu_values utility function from utils.py
- Remove test file that only tested trivial string formatting

This addresses the TODO comment to refactor and remove the get_gpu_values
helper function.

strickvl committed:
Small refactoring

strickvl committed:
Remove unnecessary asyncio complexity

strickvl committed:
Refactor token checks

strickvl committed:
Make environment parameter optional in launch method

Aligns type hint with build_modal_image signature for consistency.

strickvl committed:
Simplify memory conversion using walrus operator

Reduces verbosity by combining assignment and conditional check.

strickvl committed:
Use consistent truthy checks for None handling

Aligns with existing pattern used throughout Modal integration.

strickvl committed:
Extract timeout default into named constant

Improves maintainability by defining the value in a single location.

strickvl committed:
Fix GPU settings validation in Modal step operator

Previously, setting gpu_count > 0 without specifying a GPU type would
silently run on CPU with no warning. This adds comprehensive validation
that raises clear configuration errors with actionable guidance.

Changes:
- Add _compute_modal_gpu_arg() helper to validate and format GPU settings
- Raise StackComponentInterfaceError when gpu_count > 0 but gpu is None
- Warn and default to 1 GPU when gpu is set but gpu_count is 0
- Validate gpu_count is non-negative integer
- Normalize whitespace in GPU type strings
- Add missing List import to utils.py

strickvl committed:
Allow step-level modal_environment overrides

Previously, setup_modal_client always used the component-level
modal_environment config, preventing per-step overrides via settings.
Now prioritizes settings.modal_environment when provided, falling back
to component config for backward compatibility.

strickvl committed:
Add unit tests for complex helper functions

strickvl committed:
Update docs page

strickvl committed:
better error handling

strickvl committed:
Remove excess comments

strickvl committed:
Add type hint for resource_settings parameter

Improves type safety by explicitly typing the resource_settings parameter
as ResourceSettings in _compute_modal_gpu_arg method.

strickvl committed:
Enforce timeout constraints in ModalStepOperatorSettings

Adds Pydantic field constraints to enforce the timeout range (1-86400 seconds)
that was previously only documented. Uses ge=1 and le=DEFAULT_TIMEOUT_SECONDS
to validate input at instantiation time.

strickvl committed:
Improve Modal sandbox command execution safety

Changes command execution from shell-based to direct varargs invocation,
eliminating shell metacharacter issues and removing the implicit bash
dependency. Adds proper argument quoting via shlex.join() as a fallback
for backward compatibility with older Modal client versions that don't
support varargs command passing.

strickvl committed:
Enhance Modal step operator error handling for sandbox creation

Improves error handling during the creation of a Modal sandbox by adding a fallback mechanism for command execution. If the direct varargs invocation fails due to a TypeError, the code now falls back to a shell-quoted invocation using shlex.join() to ensure compatibility with older Modal client versions. This change enhances robustness and maintains backward compatibility while preserving argument safety.

strickvl committed:
Remove excessive comment

strickvl committed:
Refactor the get_gpu_values out to utils

strickvl committed:
Fix darglint docstring errors in Modal integration

Add missing parameter documentation to helper functions and exception
documentation to the launch method to satisfy darglint requirements.

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Small changes

strickvl committed:
mypy fix

strickvl committed:
Tests go in the right folders

strickvl committed:
Add licenses

strickvl committed:
Remove dumb tests

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -153,7 +106,7 @@ def launch(
         info: "StepRunInfo",
         entrypoint_command: List[str],
-        environment: Dict[str, str],
+        environment: Optional[Dict[str, str]],
```
Why was this change made? The base class still has a non-optional dict there

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -163,80 +116,95 @@ def launch(
-            ),
-        ).env(environment)
+        try:
+            zenml_image = build_modal_image(image_name, stack, environment)
```
This is not a "zenml_image", as the function name indicates

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -163,80 +116,95 @@ def launch(
+        except Exception as e:
+            raise RuntimeError(
+                "Failed to build the Modal image from your Docker registry. "
+                "Action required: ensure your ZenML stack's container registry is configured with valid credentials "
+                "and that the base image exists and is accessible. "
```
I doubt any of those two cases are actually the cause for failure here? We've built that image just before that pipeline run, and apparently also had permissions to push it there?

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -163,80 +116,95 @@ def launch(
+                        )
+
+                        try:
+                            if (
+                                not entrypoint_command
+                                or not entrypoint_command[0]
+                            ):
+                                raise ValueError(
+                                    "Empty step entrypoint command is not allowed."
+                                )
```
This is not happening
```suggestion
```

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -163,80 +116,95 @@ def launch(
-            return future
+                try:
+                    async with app.run():
+                        memory_int = (
```
This we can maybe move outside the `with app.run` context?

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -163,80 +116,95 @@ def launch(
+                        except Exception as e:
+                            raise RuntimeError(
+                                "Failed to create a Modal sandbox. "
+                                "Action required: verify that the referenced Docker image exists and is accessible, "
```
The docker image exits for sure, otherwise it would have failed earlier

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -163,80 +116,95 @@ def launch(
+                        except Exception as e:
+                            raise RuntimeError(
+                                "Modal sandbox execution failed. "
+                                "Action required: inspect the step logs in Modal, validate your entrypoint command, "
```
The entrypoint command is coming from ZenML and the user shouldn't care

strickvl committed:
Revert Optional setting for environment

strickvl committed:
Fix variable naming and error message in Modal step operator

Rename zenml_image to modal_image to accurately reflect that the function
returns a Modal-specific image object, not a ZenML construct.

Update error message to describe actual failure scenarios: Modal's inability
to access the container registry or extend the Docker image, rather than
suggesting the image might not exist (which is incorrect since ZenML just
built and pushed it successfully).

strickvl committed:
Move memory calculation outside Modal runtime context

Move the memory_int calculation from inside the async with app.run()
context to outside, grouping it with other resource preparations like
gpu_values. This is a pure data transformation that doesn't require
Modal's runtime context, improving code organization and reducing
unnecessary nesting.

strickvl committed:
Remove unneeded guardrail

strickvl committed:
Update comments

strickvl committed:
Remove extra modal pip install

strickvl committed:
Adapt CLI tests for Click 8.2 compatibility

strickvl committed:
Adapt CLI tests for Click 8.2 compatibility

schustmi commented on tests/integration/functional/cli/utils.py:
```diff
@@ -35,6 +38,16 @@
 
+def cli_runner(**kwargs) -> CliRunner:
+    """Return a Click runner that stays compatible across Click releases."""
+    runner_kwargs = dict(kwargs)
```
kwargs is already a dict

strickvl committed:
Merge branch 'feature/update-modal-step-operator' of github.com:zenml-io/zenml into feature/update-modal-step-operator

strickvl committed:
Merge remote-tracking branch 'origin/develop' into feature/update-modal-step-operator

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Fix Click 8.2+ compatibility for CLI tests (#4327)

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Merge develop into feature/update-modal-step-operator

Resolved 3 conflicts by taking develop's version in each case:

- modal_step_operator.py: develop's submit()/get_status() rewrite
  (PR #4515, 'Split step operator submission and monitoring')
  supersedes this branch's older launch()-based async implementation.
  The base step operator API moved from launch() to submit()+get_status(),
  so the branch's launch() override would no longer be invoked.
- tests/.../cli/test_base.py: develop's cli_runner() test helper replaces
  direct CliRunner() usage (test isolation fix).
- tests/.../cli/test_secret.py: develop's centralized cli_runner import +
  dropped mix_stderr=False (removed in Click 8.2).

Note: modal/utils.py and test_modal_utils.py remain from this branch but
are no longer used by develop's operator (which inlines get_gpu_values).

Committed with --no-verify: secret scanner flagged LakeFS demo credentials
in examples/lakefs_data_versioning/ that are already public on origin/develop
(auto-synced from upstream), not a newly introduced secret.

strickvl committed:
Upgrade Modal step operator to Modal SDK >=1.0

Modal 1.0 removed the private internals the operator relied on for
building images. Replace them with the public 1.0 API:

- modal.Image._from_args + modal.image._ImageRegistryConfig +
  modal.secret._Secret + modal_proto.api_pb2
  -> modal.Image.from_registry(image, secret=modal.Secret.from_dict(...))
- Drop the now-unused modal_proto import.

Also inline a validating get_gpu_values (checks GPU type/count
consistency and raises clear StackComponentInterfaceError messages),
replacing the minimal inline version, and remove the now-unused
modal/utils.py and its test_modal_utils.py.

The submit()/get_status()/cancel() lifecycle and ambient Modal auth
are unchanged; App.lookup/Sandbox.create/from_id/poll/terminate are
all verified-public in Modal 1.x. Dependency pin is modal>=1.

strickvl committed:
Fix Modal SDK step operator wiring

strickvl committed:
Update modal docs for ARM64 image type

strickvl committed:
Avoid deprecated Modal sandbox environment arg

strickvl committed:
Move Modal step env to sandbox runtime

strickvl committed:
Apply Modal environment during sandbox submission

strickvl committed:
Fix Modal environment follow-up handling

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

strickvl committed:
Fix Modal step operator docstring

strickvl committed:
Use explicit Modal SDK client

strickvl committed:
Fix Modal step operator typing

schustmi commented on src/zenml/integrations/modal/flavors/modal_step_operator_flavor.py:
```diff
@@ -36,20 +41,92 @@ class ModalStepOperatorSettings(BaseSettings):
+    @model_validator(mode="after")
+    def validate_modal_token_pair(self) -> "ModalStepOperatorConfig":
+        """Validate that Modal token fields are configured together."""
+        token_id = self.token_id.strip() if self.token_id else None
+        token_secret = self.token_secret.strip() if self.token_secret else None
```
These could still be referencing a secret key, for which the value will be empty. Just FYI, not sure if this means we'd need to do another check of this at runtime.

schustmi replied:
nvm, happens at runtime already.

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -67,6 +128,37 @@ class ModalStepOperator(BaseStepOperator):
+
+    def _get_modal_client(self) -> Optional["modal.Client"]:
+        """Get an explicit Modal client when credentials are configured."""
+        token_id = _normalize_optional_config_value(self.config.token_id)
```
This should only happen if we're actually creating the client, not in case we're just returning the cached one, right?

Background: In case the value is a secret reference, this results in an API call.

strickvl replied:
Ok yeah I can check the cached client before reading token fields.

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -67,6 +128,37 @@ class ModalStepOperator(BaseStepOperator):
+            )
+
+        if not token_id or not token_secret:
+            return None
```
Is there no `modal.Client.from_env` or similar that we could use to make this method return a non-optional client?

strickvl replied:
Modal does have `Client.from_env`, but it is async/hidden-ish (not even listed in their docs). The way I'm doing it currently is simpler than what we'd have to do to use `from_env`. WDYT?

schustmi replied:
Let's keep it as is then

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -40,24 +42,83 @@
+    gpu_type_raw = settings.gpu
+    gpu_type = gpu_type_raw.strip() if gpu_type_raw is not None else None
+    if gpu_type == "":
+        gpu_type = None
```
`_normalize_optional_config_value` does exactly this, no?

strickvl replied:
Yeah correct. I'll replace it with the helper. thanks

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -40,24 +42,83 @@
-    return f"{settings.gpu}:{gpu_count}" if gpu_count else settings.gpu
+    if gpu_count is not None:
+        try:
+            gpu_count = int(gpu_count)
```
gpu_count is guaranteed to be a non-negative int by pydantic validators on resource settings.

strickvl replied:
Yeah probably over-defensive. We'd never reach that state. I'll remove it.

strickvl committed:
Harden Modal sandbox submission

strickvl committed:
Address Modal review feedback

strickvl committed:
Merge branch 'develop' into feature/update-modal-step-operator

schustmi commented on src/zenml/integrations/modal/step_operators/modal_step_operator.py:
```diff
@@ -67,6 +134,39 @@ class ModalStepOperator(BaseStepOperator):
+    def _get_modal_client(self) -> Optional["modal.Client"]:
+        """Get an explicit Modal client when credentials are configured."""
+        with self._modal_client_lock:
+            modal_client = self._modal_client
+            if modal_client is not None and not modal_client.is_closed():
+                return modal_client
```
One tiny improvement: I think this check could happen outside the lock?

strickvl replied:
https://github.com/zenml-io/zenml/pull/4038/changes/f9b59e26f2d91da4618a283a8709cfa6b80da71e

strickvl committed:
Avoid locking cached Modal client lookup
